include/hwmalloc/heap.hpp

```diff
@@ -98,78 +103,71 @@ class heap
-    , m_num_reserve_segments{num_reserve_segments}
-    , m_tiny_heaps(s_tiny_limit / s_tiny_increment)
-    , m_heaps(bucket_index(m_max_size) + 1)
+    heap(Context* context, heap_config const& config = get_default_heap_config())
+    : m_config{config}
+    , m_context{context}
+    , m_max_size(
+          std::max(detail::round_to_pow_of_2(m_config.m_large_limit * 2), m_config.m_large_limit))
+    , m_tiny_heaps(m_config.m_tiny_limit / m_config.m_tiny_increment)
+    , m_heaps(bucket_index(m_max_size, m_config.m_bucket_shift) + 1)
     {
         for (std::size_t i = 0; i < m_tiny_heaps.size(); ++i)
             m_tiny_heaps[i] = std::make_unique<fixed_size_heap_type>(m_context,
-                s_tiny_increment * (i + 1), s_tiny_segment, m_never_free, m_num_reserve_segments);
+                m_config.m_tiny_increment * (i + 1), m_config.m_tiny_segment_size,
+                m_config.m_never_free, m_config.m_num_reserve_segments);
 
-        for (std::size_t i = 0; i < s_num_small_heaps; ++i)
+        for (std::size_t i = 0; i < m_config.m_num_small_heaps; ++i)
             m_heaps[i] = std::make_unique<fixed_size_heap_type>(m_context,
-                (s_tiny_limit << (i + 1)), s_small_segment, m_never_free, m_num_reserve_segments);
+                (m_config.m_tiny_limit << (i + 1)), m_config.m_small_segment_size,
+                m_config.m_never_free, m_config.m_num_reserve_segments);
+
+        for (std::size_t i = 0; i < m_config.m_num_large_heaps; ++i)
+            m_heaps[i + m_config.m_num_small_heaps] = std::make_unique<fixed_size_heap_type>(
+                m_context, (m_config.m_small_limit << (i + 1)), m_config.m_large_segment_size,
+                m_config.m_never_free, m_config.m_num_reserve_segments);
 
-        for (std::size_t i = 0; i < s_num_large_heaps; ++i)
-            m_heaps[i + s_num_small_heaps] = std::make_unique<fixed_size_heap_type>(m_context,
-                (s_small_limit << (i + 1)), s_large_segment, m_never_free, m_num_reserve_segments);
+        for (std::size_t i = 0;
+             i < m_heaps.size() - (m_config.m_num_small_heaps + m_config.m_num_large_heaps); ++i)
+            m_heaps[i + m_config.m_num_small_heaps + m_config.m_num_large_heaps] =
+                std::make_unique<fixed_size_heap_type>(m_context,
+                    (m_config.m_large_limit << (i + 1)), (m_config.m_large_limit << (i + 1)),
+                    m_config.m_never_free, m_config.m_num_reserve_segments);
 
-        for (std::size_t i = 0; i < m_heaps.size() - (s_num_small_heaps + s_num_large_heaps); ++i)
-            m_heaps[i + s_num_small_heaps + s_num_large_heaps] =
-                std::make_unique<fixed_size_heap_type>(m_context, (s_large_limit << (i + 1)),
-                    (s_large_limit << (i + 1)), m_never_free, m_num_reserve_segments);
+        if (m_config.m_num_huge_alloc_warn_threshold > 0)
+            m_huge_segment_alloc_cb = [this] () {
+              m_num_huge_alloc += 1;
```

Comment:
Open question that we didn't find a final answer for: Does this need to be atomic since this may be incremented concurrently for different heaps?

I think our conclusion was that each pool is thread safe, but e.g. updating a map of pools is not under a lock, so the assumption is that at this level allocations are anyway serialized? Thus no need for an atomic?